components/templates/src/global_fns/mod.rs

```diff
@@ -74,11 +74,35 @@ fn make_path_with_lang(path: String, lang: &str, config: &Config) -> Result<Stri
     Ok(splitted_path.join("."))
 }
 
-fn compute_file_sha256(path: &PathBuf) -> result::Result<String, io::Error> {
-    let mut file = fs::File::open(path)?;
+fn open_file(search_paths: &Vec<PathBuf>, url: &String) -> result::Result<fs::File, String> {
+    let cleaned_url = url.trim_start_matches("@/").trim_start_matches("/");
+    for base_path in search_paths {
+        match fs::File::open(base_path.join(cleaned_url)) {
+            Ok(f) => return Ok(f),
+            _ => continue
+        };
+    }
+    Err(format!("file `{}` not found; searched in{}", url,
+        search_paths.iter().fold(String::new(), |acc, arg| acc + " " + arg.to_str().unwrap())))
+}
+
+fn compute_file_sha256(mut file: fs::File) -> result::Result<String, String> {
     let mut hasher = Sha256::new();
-    io::copy(&mut file, &mut hasher)?;
-    Ok(format!("{:x}", hasher.result()))
+    io::copy(&mut file, &mut hasher)
+        .and_then(|_| Ok(format!("{:x}", hasher.result())))
+        .map_err(|e| format!("{}", e))
+}
+fn compute_file_sha384(mut file: fs::File) -> result::Result<String, String> {
+    let mut hasher = Sha384::new();
+    io::copy(&mut file, &mut hasher)
+        .and_then(|_| Ok(format!("{:x}", hasher.result())))
+        .map_err(|e| format!("{}", e))
+}
+fn compute_file_sha512(mut file: fs::File) -> result::Result<String, String> {
```

Comment:
Can we not return an Error rather than a string so we can use `?`?

Reply:
I'll see what I can do. I had some difficulties getting everything to work the way I want, but that's just because I'm a Rust newbie. :)

Reply:
Yeah, this is better. I split the error message creation into its own function.